### PR TITLE
Replace OGC with gcloud CLI for integration tests

### PR DESCRIPTION
**Pull request by blakerouse**
## What does this PR do?

Removes the usage of OGC from the integration runner for local development.

## Why is it important?

Enables the ability to test arm64 machines locally (OGC didn't support arm64). Removes a dependency and the `gcloud` CLI was already required for authentication.

## Checklist

- [x] I have read and understood the [pull request guidelines](https://github.com/elastic/elastic-agent/blob/main/CONTRIBUTING.md#pull-request-guidelines) of this project.
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- ~~[ ] I have added tests that prove my fix is effective or that my feature works~~
- ~~[ ] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)~~
- ~~[ ] I have added an integration test or an E2E test~~

## Disruptive User Impact

None

## How to test this PR locally

Run `mage integration:test` or `mage integration:single` to now have it use `gcloud`.

### CODE REVIEW

**Comment by blakerouse** [comment]
Failing with FIPS stack cleanup, which is not related to this PR. Retrying didn't fix that.

**Comment by adam-stokes** [comment]
We've sunsetted ogc in favor of oblt-cli https://github.com/elastic/ogc/issues/5 just a heads up

**Comment by fr4nc1sc0-r4m0n** [comment]
> We've sunsetted ogc in favor of oblt-cli https://github.com/elastic/ogc/issues/5 just a heads up

As @adam-stokes mentions, I think using the `oblt-cli` will be the right approach here for instance in terms of security since it uses Ephemeral Tokens and there's no need for a SA among others.

What do you think?

**Comment by blakerouse** [comment]
We already had `gcloud` in the `integration:auth` so I would prefer to just use the `gcloud` CLI just simpler and easier. No need to add another dependency for a developer to run this on there machine. It sets all the correct labels, so cleanup will be handled from that.

**Comment by v1v** [comment]
> No need to add another dependency for a developer to run this on there machine.

IIRC, we don't use anymore ogc in the CI but specific VMs with the required tools, is that right?

> It sets all the correct labels, so cleanup will be handled from that.

If it runs in the CI can you add `ephemeral: true`? That's the one we use to run some last resort cleanup process in our end.

**Comment by blakerouse** [comment]
> > No need to add another dependency for a developer to run this on there machine.
> 
> IIRC, we don't use anymore ogc in the CI but specific VMs with the required tools, is that right?
> 
> > It sets all the correct labels, so cleanup will be handled from that.
> 
> If it runs in the CI can you add `ephemeral: true`? That's the one we use to run some last resort cleanup process in our end.

This is not a CI change, this is just for developers on there machine.

**Comment by ycombinator** [comment]
Ran an integration test locally using `mage -v integration:single` and verified that gcloud is being used for provisioning 👍.  Reviewing code now...

File: .buildkite/misc/gce-cleanup.yml [hunk 1]
```diff
@@ -27,7 +27,7 @@ scanners:
             param: "elastic-agent"
           - type: "regex"
             pointer: "/name"
-            param: "^ogc(.*)"
+            param: "^(ogc|linux-|vm-|windows-)(.*)"
```
Comment (ycombinator) [inline .buildkite/misc/gce-cleanup.yml]: Nit: Maybe we could have `sanitizeInstanceName` return instance names with a short prefix like `ea-` or `eat-` and simplify this regex?

File: pkg/testing/gcloud/provisioner.go [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+		"--boot-disk-size", "50GB",
+		"--labels", strings.Join(labelParts, ","),
+		"--format", "json",
+		"--quiet",
+	}
+
+	if batch.OS.Type == define.Windows {
+		// Windows needs a startup script to install OpenSSH and configure the public key.
+		// The enable-windows-ssh metadata alone is not sufficient for key-based auth.
+		startupScript := windowsStartupScript(layout.Username, strings.TrimSpace(publicKey))
```
Comment (ycombinator) [inline pkg/testing/gcloud/provisioner.go]: Nit: `strings.TrimSpace(publicKey)` is being called here and earlier too. Maybe extract it to a variable or constant?

File: pkg/testing/gcloud/provisioner.go [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+# Restart sshd to pick up configuration changes
+Restart-Service sshd
+`, username, publicKey, publicKey)
+}
+
+// sanitizeInstanceName converts a batch ID into a valid GCE instance name.
+// GCE names must be lowercase, start with a letter, contain only letters/numbers/hyphens, max 63 chars.
+var invalidGCEChars = regexp.MustCompile(`[^a-z0-9-]`)
+
+func sanitizeInstanceName(batchID string) string {
```
Comment (ycombinator) [inline pkg/testing/gcloud/provisioner.go:357]: Would be nice to have some unit tests for this function, particularly around the edge cases of 63 character limit and invalid characters.

File: .buildkite/misc/gce-cleanup.yml (hunk 1 above)
Comment (ycombinator) [reply .buildkite/misc/gce-cleanup.yml]: Or we could drop this filter altogether, add a new label-based filter for a label like `purpose: integration-test`, and have that be set in the `instanceLabels` map in `pkg/testing/gcloud/provisioner.go`.  That might be more robust than depending on an instance name regex?

File: pkg/testing/gcloud/provisioner.go (hunk 3 above)
Comment (blakerouse) [reply pkg/testing/gcloud/provisioner.go:357]: Added.

File: pkg/testing/gcloud/provisioner.go (hunk 2 above)
Comment (blakerouse) [reply pkg/testing/gcloud/provisioner.go]: Moved to a variables.

File: .buildkite/misc/gce-cleanup.yml (hunk 1 above)
Comment (blakerouse) [reply .buildkite/misc/gce-cleanup.yml]: Went with the purpose label and filter.

File: pkg/testing/gcloud/provisioner.go [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+			}
+		}
+	}
+	return ""
+}
+
+// windowsStartupScript returns a PowerShell script that installs and configures
+// OpenSSH server on a Windows GCE instance, including adding the provided SSH
+// public key for the given user and setting the default shell to PowerShell.
+func windowsStartupScript(username, publicKey string) string {
```
[the remark this answers is not in the document]
Comment (blakerouse) [reply pkg/testing/gcloud/provisioner.go:303]: Fixed.

File: pkg/testing/gcloud/provisioner.go [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+		process.WithCmdOptions(runner.AttachOut(&stdout), runner.AttachErr(&stderr)),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to start gcloud %s: %w", args[0], err)
+	}
+	ps := <-proc.Wait()
+	if ps.ExitCode() != 0 {
+		return nil, fmt.Errorf("gcloud %s exited with code %d: %s", strings.Join(args, " "), ps.ExitCode(), stderr.String())
+	}
+	return stdout.Bytes(), nil
```
[the remark this answers is not in the document]
Comment (blakerouse) [reply pkg/testing/gcloud/provisioner.go:265]: Fixed.

File: pkg/testing/gcloud/provisioner.go [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+	var stderr bytes.Buffer
+	proc, err := process.Start("gcloud",
+		process.WithContext(ctx),
+		process.WithArgs(args),
+		process.WithCmdOptions(runner.AttachOut(&stdout), runner.AttachErr(&stderr)),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to start gcloud %s: %w", args[0], err)
+	}
+	ps := <-proc.Wait()
```
[the remark this answers is not in the document]
Comment (blakerouse) [reply pkg/testing/gcloud/provisioner.go:255]: Fixed.

**Review by ycombinator — DISMISSED** [review]

**Comment by v1v** [comment]
> This is not a CI change, this is just for developers on there machine.

I understand, but if those VMs are not destroyed, who will delete them? 
We need to ensure we don't have any idling resources doing nothing but costing money; that's why we built a mechanism to delete any VMs older than 24 hours with the label `ephemeral: true`, regardless of whether they run in the CI or by developers.

Labels are the only way to know when we can delete things.

**Comment by blakerouse** [comment]
@v1v I added the `ephemeral` label.

@ycombinator Can you give this another review, being adding that label dismissed the review.

**Review by v1v — APPROVED** [review]

**Comment by blakerouse** [comment]
buildkite test this

**Review by ycombinator — APPROVED** [review]
LGTM.  Thanks for adding the additional unit tests and switching to the `purpose` label.
